### ISSUE TITLE
Assistant2 CORS policy error. Access to XMLHttpRequest: No 'Access-Control-Allow-Origin' header.

### ISSUE THREAD

**Issue by npacucci** [watson-developer-cloud/node-sdk#884]
Hi, 
I'm trying to use the Assistant2 APIs inside my Web Application (Angular CLI project) with the suggested [Webpack Configuration ](https://github.com/watson-developer-cloud/node-sdk/tree/master/examples/webpack).

I followed the official docs to use the APIs on pure Client scenario, as described here:
- https://cloud.ibm.com/docs/services/assistant?topic=assistant-api-client&programming_language=javascript&locale=en
- https://cloud.ibm.com/apidocs/assistant-v2?locale=it&code=node#authentication

**Overview**

I want to authenticate by _iam_apikey_, so my configuration is something like this:

```
import * as Assistant2 from 'ibm-watson/assistant/v2';

const assistant = new Assistant2({
  version: '2019-02-28',
  iam_apikey: '****************',
  url: 'https://gateway-lon.watsonplatform.net/assistant/api'
});

assistant.createSession({
    assistant_id: '***************'
})
.then(res => {
    console.log(JSON.stringify(res, null, 2));
 })
 .catch(err => {
   console.log(err);
 });
```

Debugging at **runtime**, after the Assistant2 initialization and before the _createSession_ method execution, the assistant object looks like this:

```
assistant :  {
     tokenManager: {
          iamApikey: '****************',
          iamUrl: 'https://iam.cloud.ibm.com/identity/token',
          tokenInfo: { }
      },
     _options: {
          iam_apikey: '****************',
          qs: {
                version: '2019-02-28'
           },
          rejectUnauthorized: true,
          url: 'https://gateway-lon.watsonplatform.net/assistant/api',
          version: '2019-02-28'
     }
}
```

But executing the _createSession_ method, I get the following errors:

```
Refused to set unsafe header "Accept-Encoding" (xhr.js:126).

Access to XMLHttpRequest at 'https://iam.cloud.ibm.com/identity/token' from origin 'http://localhost:5001' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

**It seems that the IAM server that should generate the access token is not working.** 
As quoted [here](https://github.com/watson-developer-cloud/node-sdk/issues/865#issuecomment-488436479) by @germanattanasio . 

**Expected behavior**
The Assistant2 support CORS so I expected to simply call the APIs authenticating by the _iam_apikey_.

**Actual behavior**
CORS policy error.

**SDK Version**
ibm-watson: 4.1.1

**Comment by notarnet** [comment]
Any news? @germanattanasio @dpopp07

**Comment by npacucci** [comment]
Hi @germanattanasio, I tried getting the access token server side by an api in my app (as you described [here ](https://github.com/watson-developer-cloud/node-sdk/issues/865#issuecomment-488436479)) and it works. 

But after I set the access token inside the assistant2 constructor in my client code (typescript):

```
this.assistant = new Assistant({
                        version: this.version,
                        iam_access_token: <GENERATED_TOKEN>,
                        url: this.url
                    });
```

and executing the _createSession_ method I get another CORS error calling the gateway-lon.watsonplatform url:

`Access to XMLHttpRequest at 'https://gateway-lon.watsonplatform.net/assistant/api/v2/assistants/23********65/sessions?version=2019-02-28' from origin 'http://localhost:5001' has been blocked by CORS policy: Request header field x-ibmcloud-sdk-analytics is not allowed by Access-Control-Allow-Headers in preflight response.                    
`

**So maybe I misunderstood. What do you mean when you say that the assistant supports CORS?
Do you expect that it is possible to make API calls via browser or calls must be all made server side to avoid the CORS policy errors?**

**Comment by germanattanasio** [comment]
Ohhh That's a problem on our side @npacucci. Let me work on it and get back to you.

**Comment by germanattanasio** [comment]
We have a PR to fix this in our servers

**Comment by npacucci** [comment]
@germanattanasio, I tried using the PR #776 but I couldn't do it. That PR is too old because it was branched from the older master _'watson-developer-cloud'_, so most code has been refactored now inside the new _'ibm-cloud-sdk-core'_ sdk and than the Assistant2 didn't even exist. 

Is there any news about this fix to your servers?

**Comment by germanattanasio** [comment]
Changes are being tested in `dev` I expect them to be in `staging` next week and hopefully prod a day or two after staging

**Comment by dpopp07** [comment]
@germanattanasio @npacucci What is the status of this issue?

**Comment by germanattanasio** [comment]
Still in dev. Something with ingress came up so they are working on that first

**Comment by dpopp07** [comment]
@npacucci Thinking about this more, there may be a workaround you can use until things are fixed on our side. If the problem is coming from the request headers, you should be able to [configure `axios`](https://github.com/axios/axios#request-config), our internal request library, to access the request object before it is sent and remove the headers.

We expose most of the `axios` config parameters in the service constructor so that the user has some control over their requests. I think the `transformRequest` parameter would suit your needs. Their documentation for that parameter even says
> You may modify the headers object.

Let me know if you are able to use that workaround and if it works for you.

**Comment by germanattanasio** [comment]
The changes to support CORS are now in production 🎉

**Comment by drwhirst** [comment]
Sill doesn't work for me

**Comment by drwhirst** [comment]
Access to XMLHttpRequest at 'https://iam.cloud.ibm.com/identity/token' from origin 'http://localhost:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.

**Comment by drwhirst** [comment]
@germanattanasio

**Comment by germanattanasio** [comment]
The node-sdk can be used in a browser but not the IAM service. You are getting an error because you are trying to use your `apikey` in the constructor and the SDK is trying to get a token with it.

The recommendation is to have that interaction between <apikey>  and <access_token> in the server-side. 

The node-sdk will then use the `access_token`.

This is a security feature. The fact that the node-sdk accepts an apikey is just for convenience since we do the token exchange behind scenes

**Comment by npacucci** [comment]
> The node-sdk can be used in a browser but not the IAM service. You are getting an error because you are trying to use your `apikey` in the constructor and the SDK is trying to get a token with it.
> 
> The recommendation is to have that interaction between `<apikey>` and `<access_token>` in the server-side.
> 
> The node-sdk will then use the `access_token`.
> 
> This is a security feature. The fact that the node-sdk accepts an apikey is just for convenience since we do the token exchange behind scenes

Hi @germanattanasio , I tried your solution and ideed it works now, but only on Chrome or Opera browsers.

In Edge, Mozilla Firefox or Safari cases I get the following error using the **assistant.message()** method:

> SEC7123: [CORS] The origin 'http://localhost:5001' blocked a preflight request where Request Header 'User-Agent' was not present in the Access-Control-Allow-Headers response header list for a cross-origin request to 'https://gateway-lon.watsonplatform.net/assistant/api/v2/assistants/*****************/sessions?version=2019-02-28'.

Now the problem in your server is on the "Access-Control-Allow-Headers" header, because it doesn't return the 'user-agent' value on  **response**.

**Edge, Firefox or Safari scenario:**

The **request** header has:

`Access-Control-Request-Headers: authorization,user-agent,x-ibmcloud-sdk-analytics,x-watson-learning-opt-out`

And the **response** header has all of the previous requested values plus others, but not the 'user-agent' value:

`Access-Control-Allow-Headers: Content-Type, Content-Length, Authorization, X-Watson-Authorization-Token, X-WDC-PL-OPT-OUT, X-Watson-UserInfo, X-Watson-Learning-Opt-Out, X-Watson-Metadata, X-IBMCloud-SDK-Analytics`


Instead, on Chrome scenario it works luckily due to another problem inside the 'xhr.js' file, in fact the 'User-Agent' header can't be added (I don't no why) so the **request** header hasn't that value, and therefore does not expect it as a value in Access-Control-Request-Header **response**:

The **request** header in Chrome/Opera is like:

`Access-Control-Request-Headers: authorization,x-ibmcloud-sdk-analytics,x-watson-learning-opt-out`

Obviously the Access-Control-Request-Headers **response** is the same as the Edge/Firefox scenario.

So in Chrome the service works perfectly but I get the following console error:

> xhr.js:108 Refused to set unsafe header "User-Agent"


In conclusion, to make the service work on all browsers, you should just add the 'user-agent' value on the Access-Control-Allow-Headers **response**, like this:

`Access-Control-Allow-Headers: User-Agent, Content-Type, Content-Length, Authorization, X-Watson-Authorization-Token, X-WDC-PL-OPT-OUT, X-Watson-UserInfo, X-Watson-Learning-Opt-Out, X-Watson-Metadata, X-IBMCloud-SDK-Analytics`

In doing so, would remain only the "xhr.js:108 Refused to set unsafe header "User-Agent" error on Chrome, but that's is an other minor problem.

I hope I've been clear, ask me if you have doubts. If you want, I can open another ticket.

**Comment by germanattanasio** [comment]
@npacucci can you try the following workaround?

```js
const assistant = new AssistantV1({
  iam_access_token: '<access_token>',
  version: '2019-01-01',
  transformRequest: [(data, headers) => {
    delete headers['User-Agent']; // or delete headers.common['User-Agent']
    return data
  }],
});

**Comment by npacucci** [comment]
@germanattanasio the "transformRequest" parameter is not exposed on the Assistant **V2** constructor, so I don't have any entry point to delete that header.. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by dpopp07** [comment]
@npacucci That parameter is not documented but _should_ be exposed on the Assistant V2 constructor. Are you using TypeScript or a framework that won't let you add an undocumented parameter to the constructor?

**Comment by npacucci** [comment]
@dpopp07 Yes I'm using TypeScript. I'll wait the fix. 
Meanwhile I can try doing a spell to the TypeScript compiler :-)

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by npacucci** [comment]
Hi guys, the spell works fine as workaround!
I had to cast the **Options** constructor object to _any_, and also serialize the input **data**.

Like this:
```
this.assistant = new Assistant({
                        version: this.version,
                        url: this.url,
                        iam_access_token: accessToken,
                        transformRequest: [(data, headers) => {
                            //Bugfix User-Agent Access-Control-Allow-Headers error
                            if (headers != null) {
                                delete headers['User-Agent'];
                            }
                            
                            if (data != null) {
                                return JSON.stringify(data);
                            }
                            
                            return data;
                        }]
                    } as any);
```

But now it works fine on cross browser 🤓

**Comment by dpopp07** [comment]
Glad that this works! In the short term, I think we should at least be able to add additional properties to the constructor parameters type so that you don't need to cast as `any`. I'll post that fix in the next week or so.

**Comment by dpopp07** [comment]
@npacucci #925 should solve at least that casting issue

**Comment by npacucci** [comment]
@dpopp07  All right

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Douglas-J-Johnson** [comment]
Hi, I am attempting to access some of the other Watson services (tone, natural language, personality insights), and have received the following error: **Access to XMLHttpRequest at 'https://iam.cloud.ibm.com/identity/token' from origin 'http://localhost:3001' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.**  which appears to be related to, if not the same as this error.  I don't know if this the fix 925 has been applied to the other services.  I am using the classes in version 5.4.0 of the Node.js SDK.

**Comment by dpopp07** [comment]
Are you using the SDK from a browser?

**Comment by Douglas-J-Johnson** [comment]
Dustin,
Thanks for getting back to me quickly!  I was trying to use the node SDK
with vanilla JS and browserify/watchify.  I ended up going a different
route when I got stuck with SDK by translating the curl documentation into
fetches and using a backend service to generate tokens so I have something
working right now, but I am still curious about the SDK.

Thanks,
Doug

**Comment by txm3278** [comment]
I understand from reading this that Watson IAM doesn't support CORS, but what is possible with Assistant2 on browser without that? Just creating a session throws a cors error.
